src/hotspot/share/opto/memnode.cpp

```diff
@@ -4117,7 +4121,11 @@ Node* ClearArrayNode::clear_memory(Node* ctl, Node* mem, Node* dest,
 
   int unit = BytesPerLong;
   if ((offset % unit) != 0) {
-    Node* adr = new AddPNode(dest, dest, phase->MakeConX(offset));
+    Node* base = dest;
+    if (phase->type(dest)->isa_oopptr() == nullptr) {
+      base = phase->C->top();
```

Comment:
How is this possible?  Aren't all arrays in the heap?

Reply:
It seems like "dest" is always an oop here and we don't need to check isa_oopptr().

Reply:
It is always an heap address. But when used at object creation time, it's not yet an oop. It only becomes one once it is initialized. For those `Store`s to the not yet initialized object, I don't think the base edge is needed (there can't be any safepoint until the object becomes an actual oop) and code elsewhere (`InitializeNode::capture_store()`) doesn't set the base input either.

Reply:
That makes sense as long as we can guarantee there is no safepoint.